### DISCUSSION TITLE
Renovate does not recognize available updates - invalid-value

### DISCUSSION THREAD

**Discussion by jankal** [renovatebot/renovate#28294] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I am using renovate with a Shopware project. Shopware is a PHP based E-Commerce software. The relevant dependency-manager is Composer.

As I onboarded renovate to the project, I was able to receive some update PRs through it. Now as Shopware `>6.6` was released, I expect the updates for all `shopware/*` packages to appear.

I can see from the Renovate logs that the package versions apparently cannot be read for those packages.

<img width="525" alt="image" src="https://github.com/renovatebot/renovate/assets/6324332/ec17b5d9-e1aa-4c2b-8c28-e721bc9879c3">

I think this could be because of the special versioning scheme, shopware uses:

E.g. the current version is 6.6.1.0 where as the first number (6) doesn't mean anything in terms of semver but the second number (6) is treated like a semver major version, 1 in this case is the minor version and 0 the patch version.

How can I configure Renovate to propose updates for my shopware packages?

<details><summary>Contents of <code>composer.json</code>:</summary>

```json
{
    "license": "proprietary",
    "type": "project",
    "require": {
        "php": ">=8.3.0",
        "composer-runtime-api": "^2.0",
        "frosh/robots-txt": "^0.1.0",
        "shopware/administration": "~6.5.0",
        "shopware/core": "~6.5.0",
        "shopware/elasticsearch": "~6.5.0",
        "shopware/storefront": "~6.5.0",
        "symfony/flex": "~2",
        "symfony/runtime": "^6.0"
    },
    "repositories": [
        {
            "type": "path",
            "url": "custom/plugins/*",
            "options": {
                "symlink": true
            }
        },
        {
            "type": "path",
            "url": "custom/plugins/*/packages/*",
            "options": {
                "symlink": true
            }
        },
        {
            "type": "path",
            "url": "custom/static-plugins/*",
            "options": {
                "symlink": true
            }
        }
    ],
    "minimum-stability": "stable",
    "prefer-stable": true,
    "config": {
        "allow-plugins": {
            "symfony/flex": true,
            "symfony/runtime": true
        },
        "optimize-autoloader": true,
        "sort-packages": true,
        "platform": {
            "php": "8.3"
        }
    },
    "scripts": {
        "auto-scripts": {
            "assets:install": "symfony-cmd"
        },
        "post-install-cmd": [
            "@auto-scripts"
        ],
        "post-update-cmd": [
            "@auto-scripts"
        ]
    },
    "extra": {
        "symfony": {
            "allow-contrib": true,
            "endpoint": [
                "https://raw.githubusercontent.com/shopware/recipes/flex/main/index.json",
                "flex://defaults"
            ]
        }
    },
    "require-dev": {
        "deployer/deployer": "^7.3",
        "frosh/development-helper": "^1.0",
        "frosh/tools": "^1.3",
        "shopware/dev-tools": "^1.1"
    }
}
```

</details>

<details><summary>Contents of <code>renovate.json</code>:</summary>

```json
{
    "assignees": ["jankal"],
    "dependencyDashboard": true,
    "lockFileMaintenance": { "enabled": true },
    "constraints": {
        "php": "~8.3"
    }
}
```

</details>

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Found no satisfying versions with 'composer' versioning
DEBUG: No currentVersion or lockedVersion found for shopware/administration
DEBUG: Found no satisfying versions with 'composer' versioning
DEBUG: No currentVersion or lockedVersion found for shopware/elasticsearch
DEBUG: Found no satisfying versions with 'composer' versioning
DEBUG: No currentVersion or lockedVersion found for shopware/storefront
DEBUG: Found no satisfying versions with 'composer' versioning
DEBUG: No currentVersion or lockedVersion found for shopware/core
```

</details>

**Comment by jankal** [comment]
Maybe the issue has something to do with the corresponding `composer.lock` lockfile. I have put it into a gist for review.

https://gist.github.com/jankal/166be41c02d2ede4431c4665ebcbab64

**Comment by rarkins** [comment, marked as answer]
Unfortunately Renovate's composer versioning doesn't support 4-part versions like 1.2.3.4 - it's based off semver.

Here's one of the packages you mention: https://semver.madewithlove.com/?package=shopware%2Fadministration

You could switch versioning to `regex` or `loose` for these packages but unfortunately those versioning have no concept of constraints such as `~6.5`. So you'd need to switch to loose or regex versioning and also "pin" your `composer.json` values from e.g. `~6.5` to e.g. `6.6.1.0`.

**Reply by jankal** [reply]
Thanks for the prompt response @rarkins!

How would I only pin those dependencies?

I tried the following config after pinning the `shopware/*` packages to a specific version in `composer.json` and running `composer update` to regenerate the lock file.:

```json
{
    "assignees": ["jankal"],
    "dependencyDashboard": true,
    "lockFileMaintenance": { "enabled": true },
    "packageRules": [
        {
            "matchPackageNames": ["^shopware\/"],
            "groupName": "shopware core packages",
            "matchUpdateTypes": ["pin"]
        }
    ],
    "constraints": {
        "php": "~8.3"
    }
}
```

But still get the following log output:
<img width="933" alt="image" src="https://github.com/renovatebot/renovate/assets/6324332/1f278f88-f06a-40e9-9243-d2e87a598e11">

**Reply by RahulGautamSingh** [reply]
You seem to have forgotten to change the versioning.

**Reply by jankal** [reply]
@RahulGautamSingh that's more or less the question I asked @rarkins in my latest comment.
If I change the `versioning` for the whole project, this would also affect other dependencies (or wouldn't it?).

That's why I was looking for a way to just change the `versioning` for those specific packages.

**Reply by RahulGautamSingh** [reply]
My bad. You can change the versioning for individual packages.
reference: https://docs.renovatebot.com/modules/versioning/#examples-of-versioning-overrides

In your case, [`matchPackagePatterns`](https://docs.renovatebot.com/configuration-options/#matchpackagepatterns) will be better choice instead of `matchPackageNames`, which is used in the docs example.

**Reply by jankal** [reply]
@RahulGautamSingh thanks! That's exactly what I was looking for! I just looked at the wrong place in the docs (section `packageRules`).

**Reply by rarkins** [reply]
I think you'll need to pin the versions manually, because composer versioning does not understand 4 part versions. After that, set versioning=loose for those specific packages/patterns using packageRules

**Comment by rarkins** [comment]
I created #28297 and #28298
